# Part1. Introduction to GATK and the suggested workflow to call variants with GATK software
http://gatkforums.broadinstitute.org/gatk/categories/methods
http://gatkforums.broadinstitute.org/gatk/discussion/3891/calling-variants-in-rnaseq

# Part2. Preparation work

## Download GATK software
https://www.broadinstitute.org/gatk/
user name: pjl_happy@126.com
password: pjlhappy111111
I downloaded 3.6 version.

## Creat a genome order file (/public/home/pangjunling/genome/AGPv3.27)
```
java -jar ~/tool/picard-tools-1.119/CreateSequenceDictionary.jar R=B73.AGPv3.27.fa O=B73.AGPv3.27.dict
```

## Download vcf files and sort (directory: /public/home/pangjunling/genome/AGPv3.27/vcf)
```
wget ftp://ftp.ensemblgenomes.org/pub/plants/release-27/vcf/zea_mays/zea_mays.vcf.gz
gunzip zea_mays.vcf.gz
bedtools sort -header -i zea_mays.vcf > zea_mays.reorder.vcf

# Note: the chromosome order should be the same with the order in "B73.AGPv3.27.dict" above. One should check.
``` 

## Extract INDEL info from vcf file and sort (directory: /public/home/pangjunling/genome/AGPv3.27/vcf)
```
 more zea_mays.vcf |awk '{if($0~/^#/ || $0~/insertion/ || $0~/deletion/){print $0}}' > zea_mays.indel.vcf &
 bedtools sort -header -i zea_mays.indel.vcf > zea_mays.indel.reorder.vcf

# Note: the chromosome order should be the same with the order in "B73.AGPv3.27.dict" above. One should check.
```

## Data to deal with

In "/public/home/pangjunling/5DAP_project/analysis/20160628/RNA-seq_mapping-and-fpkm.rmd", we have introduced how we process the RNA-seq reads, including mapping and filter the unique mapping reads. 
we got the bam files with unique mapping reads in the directory "/public/home/pangjunling/5DAP_project/tophat_uniq_filtered/". 

# Part3. Deal with BAM files

## Add read group for bam files
perl 1_add_rg.pl & 
The perl script "1_add_rg.pl" is written to run the following commond in parallel:
```
java -Xmx2g -jar ~/tool/picard-tools-1.119/AddOrReplaceReadGroups.jar I=tophat_uniq_filtered/*bam O=bam_add_rg/*bam ID=$name LB=$name PL=illumina PU=run SM=$name
```

## Reorder bamfiles
Sometimes mapping tools output mapping results lexicographically, but GATK call SNPs based on the karyotypic order, so we need to reorder the bam files.
perl 2_reorder_bam.pl &
The perl script "2_reorder_bam.pl" is written to run the following commond in parallel:
```
 java -Xmx2g -jar ~/tool/picard-tools-1.119/ReorderSam.jar I=bam_add_rg/*bam O=bam_reorder/*bam R=/public/home/pangjunling/genome/AGPv3.27/B73.AGPv3.27.fa
```
Note:
The "R=*.fa" parameter required reference sequence to reorder reads to match.  A sequence dictionary corresponding to the reference fasta is required.  
So, we created one with CreateSequenceDictionary.jar before this step. (see # Part2. Preparation work  ## Creat a genome order file)

## Sort bamfiles
After reorder the chromosome, we should sort the reads mapping sites accordingly (ascending) in each chromosome.
perl 3_sort_bam.pl & to run the following commond for files in paralle.
```
java -jar ~/tool/picard-tools-1.119/SortSam.jar I=bam_reorder/*bam O=bam_sort/*bam SORT_ORDER=coordinate
```
Alternatively, we could use samtools to sort the bam files with the commond "samtools sort in.bam out_sorted".

## Mark Duplicates

perl 4_mark_duplicates.pl &
The perl script "4_mark_duplicates.pl" is written to run the following commond in parallel:
```
java -Xmx4g -jar ~/tool/picard-tools-1.119/MarkDuplicates.jar I=bam_sort/*bam O=bam_mark_dup/*bam M=bam_mark_dup/*matrics CREATE_INDEX=true VALIDATION_STRINGENCY=LENIENT
```

## Split'N'Trim and reassign mapping qualities
perl 5_splitNCigarReads.pl &
This perl script is written to run the following commond in parallel:
```
java -Xmx4g -jar /public/home/pangjunling/tool/GenomeAnalysisTK-3.6/GenomeAnalysisTK.jar -T SplitNCigarReads -R /public/home/pangjunling/genome/AGPv3.27/B73.AGPv3.27.fa -I bam_mark_dup/*bam -o bam_splitN/*bam -U ALLOW_N_CIGAR_READS -fixNDN
```
Note that sometimes the commond will encounter errors with message like this "ERROR MESSAGE: Bad input: Cannot split this read (might be an empty section between Ns, for example 1N1D1N): 31M407N1D1033N19M".
To fit this, one should add "-fixNDN" parameter.

Another error one may encounter is about the quality in the bam files. Sometimes the mapping tools use different. "ERROR MESSAGE: SAM/BAM file SAMFileReader{<file.bam>} appears to be using the wrong encoding for quality scores..."  
The standard format for quality score encodings is that Q0 == ASCII 33 according to the SAM specification. However, in some datasets (including older Illumina data), encoding starts at ASCII 64. 
Two possible ways to deal with this situation: 
(1) With "--fix_misencoded_quality_scores" / "-fixMisencodedQuals". What will happen is that the engine will simply subtract 31 from every quality score as it is read in.
(2) Use "-allowPotentiallyMisencodedQuals" to override this encoding check. This is with some risk, unless you make sure the data is good.
See details: https://www.broadinstitute.org/gatk/guide/article?id=6470 

## RealignerTargetCreator
The RealignerTargetCreator step would need to be done just once for a single set of indels; so as long as the set of known indels doesn't change, the output.intervals file from below would never need to be recalculated.
See: http://gatkforums.broadinstitute.org/gatk/discussion/38/local-realignment-around-indels
```
java -Xmx20g -jar ~/tool/GenomeAnalysisTK-3.6/GenomeAnalysisTK.jar -T RealignerTargetCreator -R ~/genome/AGPv3.27/B73.AGPv3.27.fa -o zea_mays.indel.reorder.intervals -known zea_mays.indel.reorder.vcf 
```

## Indel Realigner 
See: http://gatkforums.broadinstitute.org/gatk/discussion/38/local-realignment-around-indels
perl 6_IndelRealigner.pl &
This perl script is written to run the following commond in parallel:
```
java -Xmx4g -jar /public/home/pangjunling/tool/GenomeAnalysisTK-3.6/GenomeAnalysisTK.jar -T IndelRealigner -R /public/home/pangjunling/genome/AGPv3.27/B73.AGPv3.27.fa -I bam_splitN/*bam -o bam_realign/*bam -targetIntervals /public/home/pangjunling/genome/AGPv
3.27/vcf/zea_mays.indel.reorder.intervals -known /public/home/pangjunling/genome/AGPv3.27/vcf/zea_mays.indel.reorder.vcf
```

## BaseRecalibrator
Why?
The scores produced by the machines are subject to various sources of systematic (non-random) technical error, leading to over- or under-estimated base quality scores in the data.
Base quality score recalibration (BQSR) is a process in which we apply machine learning to model these errors empirically and adjust the quality scores accordingly.
See:
https://www.broadinstitute.org/gatk/guide/tooldocs/org_broadinstitute_gatk_tools_walkers_bqsr_BaseRecalibrator.php
http://gatkforums.broadinstitute.org/gatk/discussion/44/base-quality-score-recalibration-bqsr#latest

perl 7_BaseRecalibrator.pl &
perl 8_PrintReads.pl &
This perl scripts are written to run the following commond in parallel:
```
java -Xmx4g -jar /public/home/pangjunling/tool/GenomeAnalysisTK-3.6/GenomeAnalysisTK.jar -T BaseRecalibrator  -R /public/home/pangjunling/genome/AGPv3.27/B73.AGPv3.27.fa -I bam_realign/*bam -o bam_realign/*grp -knownSites /public/home/pangjunling/genome/AGPv3.
27/vcf/zea_mays.reorder.vcf

java -Xmx4g -jar /public/home/pangjunling/tool/GenomeAnalysisTK-3.6/GenomeAnalysisTK.jar -T PrintReads -R /public/home/pangjunling/genome/AGPv3.27/B73.AGPv3.27.fa -I bam_realign/*bam -o bam_recal/*bam -BQSR bam_realign/*grp
```

# Part4. SNP calling

perl 9_snp_calling.pl & to run the following commond for all the samples in parallel:
```
java -Xmx4g -jar /public/home/pangjunling/tool/GenomeAnalysisTK-3.6/GenomeAnalysisTK.jar -T HaplotypeCaller  -R /public/home/pangjunling/genome/AGPv3.27/B73.AGPv3.27.fa -I bam_recal/*bam -dontUseSoftClippedBases -stand_call_conf 20 -stand_emit_conf 20 -o
ut_mode EMIT_ALL_SITES -o SNP_calling/*vcf
```

Two parameters to know:
--dbsnp / -D  dbSNP file
rsIDs from this file are used to populate the ID column of the output. Also, the DB INFO flag will be set when appropriate. dbSNP is not used in any way for the calculations themselves.
This argument supports reference-ordered data (ROD) files in the following formats: BCF2, VCF, VCF3

-dontUseSoftClippedBases
The developlors have added some functionality to the variant calling code which will intelligently take into account the information about intron-exon split regions that is embedded in the BAM file by SplitNCigarReads. 
In brief, the new code will perform "dangling head merging" perations and avoid using soft-clipped bases (this is a temporary solution) as necessary to minimize false positive and false negative calls. 
To invoke this new functionality, just add -dontUseSoftClippedBases to your regular HC command line. 
See: http://gatkforums.broadinstitute.org/gatk/discussion/3891/calling-variants-in-rnaseq

See:
https://www.broadinstitute.org/gatk/guide/tooldocs/org_broadinstitute_gatk_tools_walkers_haplotypecaller_HaplotypeCaller.php

# Overall notes

NOTE1: foreach step of GATK tools, it will test the scores and duplicate. So you will find a report like the following for each GATK step. 
Don't confused for the message like "8458181 reads (20.92% of total) failing DuplicateReadFilter". Because when you use picard to markduplicates, by default it will only mark the duplicates with a special flag and don't remove the duplicates. GATK will know the flag of duplicates. This will not affect the final results. 
An aternative method to deal this is to add "REMOVE_DUPLICATES=true" when running picard markDuplicates, and then for each GATK step following this, add "-drf DuplicateRead" to include duplicates in analyses, which will avoid duplicate testing.
------------------------------------
INFO  20:03:52,946 ProgressMeter - Total runtime 9176.42 secs, 152.94 min, 2.55 hours 
INFO  20:03:52,947 MicroScheduler - 8458181 reads were filtered out during the traversal out of approximately 40432006 total reads (20.92%) 
INFO  20:03:52,948 MicroScheduler -   -> 0 reads (0.00% of total) failing BadCigarFilter 
INFO  20:03:52,948 MicroScheduler -   -> 8458181 reads (20.92% of total) failing DuplicateReadFilter 
INFO  20:03:52,949 MicroScheduler -   -> 0 reads (0.00% of total) failing FailsVendorQualityCheckFilter 
INFO  20:03:52,949 MicroScheduler -   -> 0 reads (0.00% of total) failing HCMappingQualityFilter 
INFO  20:03:52,950 MicroScheduler -   -> 0 reads (0.00% of total) failing MalformedReadFilter 
INFO  20:03:52,950 MicroScheduler -   -> 0 reads (0.00% of total) failing MappingQualityUnavailableFilter 
INFO  20:03:52,951 MicroScheduler -   -> 0 reads (0.00% of total) failing NotPrimaryAlignmentFilter 
INFO  20:03:52,951 MicroScheduler -   -> 0 reads (0.00% of total) failing UnmappedReadFilter 
-------------------------------------

NOTE2: Because of the storage limites of the server, we delete the intermediate files after each step.

